File: test_data/csharp/test_main/aas_core_meta.v3rc2/input/snippets/Verification/properties_or_ranges_have_value_type.cs
/// <summary>
/// Check that the <paramref name="elements" /> which are
/// <see cref="Aas.Property" /> or <see cref="Aas.Range" />
/// have the given <paramref name="valueType" />.
/// </summary>
/// <remarks>
/// We have to use nullable valueType since the compiler does not really handle
/// nullable C# value types.
///
/// See https://endjin.com/blog/2022/02/csharp-10-generics-nullable-references-improvements-allownull
/// </remarks>
public static bool PropertiesOrRangesHaveValueType(
    IEnumerable<Aas.ISubmodelElement> elements,
    Aas.DataTypeDefXsd? valueType
)
{
    foreach (var element in elements)
    {
        switch (element)
        {
            case Aas.Property prop:
                if (prop.ValueType != valueType)
                {
                    return false;
                }
                break;
            case Aas.Range range:
                if (range.ValueType != valueType)
                {
                    return false;
                }
                break;
        }
    }
    return true;
}

File: dev/test_data/golang/test_main/aas_core_meta.v3/input/snippets/Verification/properties_or_ranges_have_value_type.go
// Check that `elements` which are [aastypes.IProperty] or [aastypes.IRange]
// have the given `valueType`.
func PropertiesOrRangesHaveValueType[E aastypes.ISubmodelElement](
	elements []E,
	valueType aastypes.DataTypeDefXSD,
) bool {
	for _, element := range elements {
		switch element.ModelType() {
		case aastypes.ModelTypeProperty:
			prop := any(element).(aastypes.IProperty)
			if prop.ValueType() != valueType {
				return false
			}
		case aastypes.ModelTypeRange:
			rng := any(element).(aastypes.IRange)
			if rng.ValueType() != valueType {
				return false
			}
		// default passes.
		}
	}
	return true
}




































